stdlib/gzip.pyi

```diff
@@ -47,17 +78,54 @@ if sys.version_info >= (3, 8):
     class BadGzipFile(OSError): ...
 
 class GzipFile(_compression.BaseStream):
-    myfileobj: Optional[IO[bytes]]
-    mode: str
+    myfileobj: Optional[FileIO]
+    mode: Literal[1, 2]
     name: str
     compress: zlib._Compress
-    fileobj: IO[bytes]
+    fileobj: Union[_ReadableFileobj, _WritableFileobj]
+    @overload
+    def __init__(
+        self,
+        filename: Optional[AnyPath],
+        mode: _ReadBinaryMode,
+        compresslevel: int = ...,
+        fileobj: Optional[_ReadableFileobj] = ...,
+        mtime: Optional[float] = ...,
+    ) -> None: ...
+    @overload
+    def __init__(
+        self,
+        *,
+        mode: _ReadBinaryMode,
+        compresslevel: int = ...,
+        fileobj: Optional[_ReadableFileobj] = ...,
+        mtime: Optional[float] = ...,
+    ) -> None: ...
+    @overload
+    def __init__(
+        self,
+        filename: Optional[AnyPath],
+        mode: _WriteBinaryMode,
+        compresslevel: int = ...,
+        fileobj: Optional[_WritableFileobj] = ...,
+        mtime: Optional[float] = ...,
+    ) -> None: ...
+    @overload
+    def __init__(
+        self,
+        *,
+        mode: _WriteBinaryMode,
+        compresslevel: int = ...,
+        fileobj: Optional[_WritableFileobj] = ...,
+        mtime: Optional[float] = ...,
+    ) -> None: ...
+    @overload
     def __init__(
         self,
         filename: Optional[AnyPath] = ...,
-        mode: Optional[str] = ...,
+        mode: Union[_ReadBinaryMode, _WriteBinaryMode, None] = ...,
         compresslevel: int = ...,
-        fileobj: Optional[IO[bytes]] = ...,
+        fileobj: Union[_ReadableFileobj, _WritableFileobj, None] = ...,
         mtime: Optional[float] = ...,
     ) -> None: ...
```

Comment:
It's possible to add more overloads to check filename/fileobj, but that could be done in a separate PR.